### DISCUSSION TITLE
Non Garbage collectable classes.

### DISCUSSION THREAD

**Discussion by IrakliLomidze** [dotnet/csharplang#5966] [Language Ideas]
In C# there is limited usage of unsafe code

But lack of some functionality sometimes it will have huge impact on performance.
Suggestion is to introduce a fixed keyword for class. Actually, I’m asking for non Garbage collectable classes. Not a just buffer.
E.g you read some entity from database, manipulate and save back.

Problems: Each time, we allocate new object in heap. During a manipulation we don’t have move semantic concept for right values as C++ does, meaning that each time new heap allocation. 

If we had a fixed class {} that we can use to read manipulation on same place and use for saving. Will fundamentally increase performance.  And finally we will be accountable to free up fixed class with our destructor.

That is a huge change, first it needs to implemented in Language level. And 2 libraries implemented in .net should support them. 

ToString()
And like ToString&() (ToFixedString())

**Comment by svick** [comment]
Why can't you use `struct`s for this?

**Reply by sharwell** [reply]
`struct` is difficult here because it's not possible to define a non-copyable struct. An instance of a reference type is known to exist in one location at any given time, so operations on the instance are visible to all other observers of that instance. If a structure gets implicitly copied to a second location, operations on one instance may or may not be visible to another.

**Reply by JustNrik** [reply]
pointers

**Reply by sharwell** [reply]
Pointers only work for types with no managed resources (e.g. a single field of type `string` will prevent a pointer to that structure).

**Reply by Unknown6656** [reply]
@sharwell : One could relax pointers to allow them to point to a managed type. The pointer value would have the size of `GCHandle` (i.e. the same size as `int`). 

Or one could at least relax pointers to point to structs containing reference type fields. The following struct
```csharp
struct A {
     int i;
     long l;
     string s;
}
```
would be handled as follows:
```csharp
struct A {
     int i;
     long l;
     GCHandle s;
}
```

**Comment by jnm2** [comment]
Similar: https://github.com/dotnet/csharplang/discussions/5542
